granulate_utils/linux/cgroups/memory_controller.py

```diff
@@ -0,0 +1,105 @@
+        return int(self.cgroup.read_from_interface_file(self.memory_limit))
+
+    def _set_swap_limit(self, limit: str) -> None:
+        try:
+            self.cgroup.write_to_interface_file(self.swap_limit, limit)
+        except PermissionError:
+            # if swap extension is not enabled (CONFIG_MEMCG_SWAP) this file doesn't exist
+            # and PermissionError is thrown (since it can't be created)
+            pass
+
+
+class MemoryControllerV1(MemoryControllerInterface):
+    memory_limit = "memory.limit_in_bytes"
+    swap_limit = "memory.memsw.limit_in_bytes"
+    max_usage_in_bytes = "memory.max_usage_in_bytes"
+    usage = "memory.usage_in_bytes"
+
+    def set_limit_in_bytes(self, limit: int) -> None:
+        # in case memsw.limit_in_bytes file exists we need to reset it in order to
+        # change limit_in_bytes in case it's smaller than memsw.limit_in_bytes
+        super()._set_swap_limit("-1")
+        self.cgroup.write_to_interface_file(self.memory_limit, str(limit))
+
+        # memsw.limit_in_bytes is already set to -1
+        if limit != -1:
+            self._set_swap_limit(str(limit))
+
+
+class MemoryControllerV2(MemoryControllerInterface):
+    UNBOUNDED_QUOTA_VALUE = "max"
+    memory_limit = "memory.max"
+    swap_limit = "memory.swap.max"
+    usage = "memory.current"
+
+    def set_limit_in_bytes(self, limit: int) -> None:
+        # In CgroupV2 swap and memory are 2 different attributes that can be tuned, as opposed to Cgroup V1,
+        # in which we can just set a limit on both swap and memory together.
+        # So in Cgroup V2 we set swap limit to be 0.
+        swap_limit = "0"
+        memory_limit = str(limit)
+        if limit == -1:
+            memory_limit = swap_limit = "max"
+        super()._set_swap_limit(swap_limit)
+        self.cgroup.write_to_interface_file(self.memory_limit, str(memory_limit))
+
+    def get_memory_limit(self) -> int:
+        memory_limit = self.cgroup.read_from_interface_file(self.memory_limit)
+        if memory_limit == self.UNBOUNDED_QUOTA_VALUE:
+            return -1
+        return int(memory_limit)
+
+
+class MemoryController(BaseController):
+    controller = "memory"
+
+    def __init__(self, cgroup: Optional[Union[Path, CgroupCore]] = None):
+        super().__init__(cgroup)
+        if isinstance(self.cgroup, CgroupCoreV1):
+            self.controller_interface: MemoryControllerInterface = MemoryControllerV1(self.cgroup)
+        elif isinstance(self.cgroup, CgroupCoreV2):
+            self.controller_interface = MemoryControllerV2(self.cgroup)
+
+    def get_memory_limit(self) -> int:
+        return self.controller_interface.get_memory_limit()
+
+    def get_usage_in_bytes(self) -> int:
+        return int(self.read_from_interface_file(self.controller_interface.usage))
+
+    def get_max_usage_in_bytes(self) -> int:
+        if self.controller_interface.max_usage_in_bytes == "":
+            raise Exception("Not implemented")
```

Comment:
If you go with my suggestion [here](https://github.com/Granulate/granulate-utils/pull/117/files#r1089821480) then this hack becomes nicely overriding `get_max_usage_in_bytes` in the v2 child, and raising an exception.

Anyway, wherever yopu end up putting the code, pls document why not implemented - I don't recall if v2 tracks this value at all? Is it supported? Did you just not implement here? Leave some breadcrumbs for the future developer adding it.

Reply:
It's not implemented for v2. I can copy the docs from the kernel at the top of each controller to show what interfaces are supported and what they mean if you think it's helpful.